tests/test_storablemap.cc

```diff
@@ -0,0 +1,161 @@
+// -*- LSST-C++ -*-
+/*
+ * This file is part of afw.
+ *
+ * Developed for the LSST Data Management System.
+ * This product includes software developed by the LSST Project
+ * (https://www.lsst.org).
+ * See the COPYRIGHT file at the top-level directory of this distribution
+ * for details of code ownership.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#define BOOST_TEST_DYN_LINK
+#define BOOST_TEST_MODULE StorableMapCpp
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-variable"
+#include "boost/test/unit_test.hpp"
+#pragma clang diagnostic pop
+
+#include <memory>
+#include <numeric>
+#include <set>
+#include <sstream>
+#include <string>
+
+#include "lsst/pex/exceptions.h"
+#include "lsst/utils/tests.h"
+
+#include "lsst/afw/typehandling/Key.h"
+#include "lsst/afw/image/detail/StorableMap.h"
+
+using namespace std;
+using namespace string_literals;
+
+namespace lsst {
+namespace afw {
+namespace image {
+namespace detail {
+
+namespace {
+class SimpleStorable : public typehandling::Storable {
+public:
+    virtual ~SimpleStorable() = default;
+
+    shared_ptr<typehandling::Storable> cloneStorable() const override {
+        return make_unique<SimpleStorable>();
+    }
+
+    string toString() const override { return "Simplest possible representation"; }
+
+    bool equals(Storable const& other) const noexcept override { return singleClassEquals(*this, other); }
+    virtual bool operator==(SimpleStorable const& other) const { return true; }
+    bool operator!=(SimpleStorable const& other) const { return !(*this == other); }
+};
+
+class ComplexStorable final : public SimpleStorable {
+public:
+    constexpr ComplexStorable(double storage) : SimpleStorable(), storage(storage) {}
+
+    ComplexStorable& operator=(double newValue) {
+        storage = newValue;
+        return *this;
+    }
+
+    shared_ptr<typehandling::Storable> cloneStorable() const override {
+        return make_unique<ComplexStorable>(storage);
```

Comment:
That's mostly me being stubborn. Methods that always create a new object should return `unique_ptr` for maximum flexibility (see Effective Modern C++; I don't have it with me so I can't quote an item number). However, `cloneStorable` can't return a `unique_ptr`, because that leads to a messy pybind11 bug that its developers aren't keen on fixing. When I write a `cloneStorable` I still create a `unique_ptr` internally because I can, and the return statement does an automatic conversion.

Reply:
Thanks for the clarification. This the kind of C++ minutiae that I wish I knew more of.